### PR TITLE
lightning: support disable scheduler by key range

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/35148

Problem Summary:

We want to use lightning to import data to an online cluster. To reduce the impact on online services, lightning shouldn't disable the whole scheduler during importing. Instead, lightning should only disable the scheduler on the region that is being imported.

### What is changed and how it works?

https://github.com/tikv/pd/pull/4695 supported ttl for region label. We can use the region label to disable scheduler on specific key range, and recover scheduler automatically.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/pdutil/pd.go [hunk 1]
```diff
@@ -700,6 +706,95 @@ func (p *PdController) doRemoveSchedulersWith(
...
+	}
+
+	go func() {
+		ticker := time.NewTicker(pauseTimeout / 3)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				if err := p.CreateOrUpdateRegionLabelRule(ctx, rule); err != nil {
+					log.Warn("pause scheduler by failed, ignore it and wait next time pause", zap.Error(err))
```
Comment (gozssky) [inline br/pkg/pdutil/pd.go:783]:
```suggestion
					log.Warn("pause scheduler by key range failed, ignore it and wait next time pause", zap.Error(err))
```

File: br/pkg/lightning/backend/local/local.go [hunk 2]
```diff
@@ -1357,6 +1357,27 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID, regi
 		return err
 	}
 
+	if len(ranges) > 0 && local.pdCtl.CanPauseSchedulerByKeyRange() {
```
Comment (glorv) [inline br/pkg/lightning/backend/local/local.go:1360]: What if several engines with range overlapping start importing at the same time, is the new scheduler compatible with this kind of concurrent operations?

Comment (gozssky) [reply br/pkg/lightning/backend/local/local.go:1360]: It's compatible. Different engines will use different label rule id. It guarantees the scheduler is disabled on every engine key range even if overlapping exists.

File: br/pkg/lightning/backend/local/localhelper.go [hunk 3]
```diff
@@ -476,12 +479,6 @@ func (local *local) checkScatterRegionFinishedOrReScatter(ctx context.Context, r
 		if respErr.GetType() == pdpb.ErrorType_REGION_NOT_FOUND {
 			return true, nil
 		}
-		// don't return error if region replicate not complete
-		// TODO: should add a new error type to avoid this check by string matching
-		matches, _ := regexp.MatchString("region \\d+ is not fully replicated", respErr.Message)
-		if matches {
-			return false, nil
-		}
 		return false, errors.Errorf("get operator error: %s", respErr.GetType())
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/localhelper.go:482]: should include `respErr.Message` in the returned error

**Review by D3Hunter — COMMENTED** [review]
will review later

File: br/pkg/lightning/backend/local/localhelper.go (hunk 3 above)
Comment (gozssky) [reply br/pkg/lightning/backend/local/localhelper.go]: Fixed in 42bf379b79afc02648a84b44664e295c56aef58a.

File: br/pkg/pdutil/pd.go [hunk 4]
```diff
@@ -709,6 +715,121 @@ func (p *PdController) doRemoveSchedulersWith(
...
+	}
+	var lastErr error
+	for i, addr := range p.addrs {
+		_, lastErr = pdRequest(ctx, addr, regionLabelPrefix,
+			p.cli, http.MethodPost, bytes.NewBuffer(reqData))
+		if lastErr == nil {
+			return nil
+		}
+		if ctx.Err() != nil {
+			return errors.Cause(ctx.Err())
```
[the remark this answers is not in the document]
Comment (gozssky) [reply br/pkg/pdutil/pd.go]: Done.

File: br/pkg/pdutil/pd.go [hunk 5]
```diff
@@ -709,6 +715,122 @@ func (p *PdController) doRemoveSchedulersWith(
...
+
+// KeyRangeRule contains the start key and end key of the LabelRule. This struct is partially copied from
+// https://github.com/tikv/pd/blob/783d060861cef37c38cbdcab9777fe95c17907fe/server/schedule/labeler/rules.go#L62.
+type KeyRangeRule struct {
+	StartKeyHex string `json:"start_key"` // hex format start key, for marshal/unmarshal
+	EndKeyHex   string `json:"end_key"`   // hex format end key, for marshal/unmarshal
+}
+
+// CreateOrUpdateRegionLabelRule creates or updates a region label rule.
+func (p *PdController) CreateOrUpdateRegionLabelRule(ctx context.Context, rule LabelRule) error {
```
Comment (D3Hunter) [inline br/pkg/pdutil/pd.go:744]: maybe add an integration test to make sure it works after rate limiter pr merged

**Review by D3Hunter — APPROVED** [review]

**Comment by gozssky** [comment]
Wait for https://github.com/tikv/tikv/pull/12858 to be merged.

File: br/pkg/lightning/backend/local/localhelper.go [hunk 6]
```diff
@@ -349,7 +341,8 @@ func (local *local) SplitAndScatterRegionByRanges(
 			zap.Int("skipped_keys", skippedKeys),
 			zap.Int("scatterCount", scatterCount),
 			zap.Int("regions", len(scatterRegions)),
-			zap.Duration("take", time.Since(startTime)))
+			zap.Duration("take", time.Since(startTime)),
+			zap.Error(err))
 	}
 	return nil
```
Comment (lichunzhu) [inline br/pkg/lightning/backend/local/localhelper.go:347]: Do we need to handle this error?

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: br/pkg/lightning/backend/local/localhelper.go (hunk 6 above)
Comment (gozssky) [reply br/pkg/lightning/backend/local/localhelper.go:347]: Emm, it's hard to say. Until now, we didn't treat scatter region error as a critical error.

File: br/pkg/lightning/backend/local/localhelper.go [hunk 7]
```diff
@@ -478,13 +481,9 @@ func (local *local) checkScatterRegionFinishedOrReScatter(ctx context.Context, r
 		if respErr.GetType() == pdpb.ErrorType_REGION_NOT_FOUND {
 			return true, nil
 		}
-		// don't return error if region replicate not complete
```
Comment (buchuitoudegou) [inline br/pkg/lightning/backend/local/localhelper.go:481]: We don't need to handle this error now, do we?

File: br/pkg/pdutil/pd.go [hunk 8]
```diff
@@ -709,6 +715,142 @@ func (p *PdController) doRemoveSchedulersWith(
...
+	go func() {
+		defer close(done)
+		ticker := time.NewTicker(ttl / 3)
+		defer ticker.Stop()
+	loop:
+		for {
+			select {
+			case <-ticker.C:
+				if err := p.CreateOrUpdateRegionLabelRule(ctx, rule); err != nil {
+					if err := errors.Cause(err); err == context.Canceled || err == context.DeadlineExceeded {
```
Comment (buchuitoudegou) [inline br/pkg/pdutil/pd.go:826]:
```suggestion
					if err := errors.Cause(err); IsContextCanceled(err) {
```
Should we reuse `IsContextCanceled` function here?

**Review by buchuitoudegou — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/backend/local/localhelper.go (hunk 7 above)
Comment (gozssky) [reply br/pkg/lightning/backend/local/localhelper.go:481]: It is moved to retriable errors.

Comment (buchuitoudegou) [reply br/pkg/lightning/backend/local/localhelper.go:481]: oh, it's caught by the caller.

**Review by buchuitoudegou — APPROVED** [review]
lgtm

**Pull request merged**